File: library-extensions-expandable/src/main/java/com/mikepenz/fastadapter/utils/AdapterUtil.java
package com.mikepenz.fastadapter.utils;

import com.mikepenz.fastadapter.FastAdapter;
import com.mikepenz.fastadapter.IExpandable;
import com.mikepenz.fastadapter.IItem;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by mikepenz on 31.12.15.
 */
public class AdapterUtil {
    /**
     * internal method to restore the selection state of subItems
     *
     * @param item          the parent item
     * @param selectedItems the list of selectedItems from the savedInstanceState
     */
    public static <Item extends IItem> void restoreSubItemSelectionStatesForAlternativeStateManagement(Item item, List<String> selectedItems) {
        if (item instanceof IExpandable && !((IExpandable) item).isExpanded() && ((IExpandable) item).getSubItems() != null) {
            List<Item> subItems = (List<Item>) ((IExpandable<Item, ?>) item).getSubItems();
            for (int i = 0, size = subItems.size(); i < size; i++) {
                Item subItem = subItems.get(i);
                String id = String.valueOf(subItem.getIdentifier());
                if (selectedItems != null && selectedItems.contains(id)) {
                    subItem.withSetSelected(true);
                }
                restoreSubItemSelectionStatesForAlternativeStateManagement(subItem, selectedItems);
            }
        }
    }

    /**
     * internal method to find all selections from subItems and sub sub items so we can save those inside our savedInstanceState
     *
     * @param item       the parent item
     * @param selections the ArrayList which will be stored in the savedInstanceState
     */
    public static <Item extends IItem> void findSubItemSelections(Item item, List<String> selections) {
        if (item instanceof IExpandable && !((IExpandable) item).isExpanded() && ((IExpandable) item).getSubItems() != null) {
            List<Item> subItems = (List<Item>) ((IExpandable<Item, ?>) item).getSubItems();
            for (int i = 0, size = subItems.size(); i < size; i++) {
                Item subItem = subItems.get(i);
                String id = String.valueOf(subItem.getIdentifier());
                if (subItem.isSelected()) {
                    selections.add(id);
                }
                findSubItemSelections(subItem, selections);
            }
        }
    }

    /**
     * Gets all items (including sub items) from the FastAdapter
     *
     * @param fastAdapter the FastAdapter
     * @return a list of all items including the whole subItem hirachy
     */
    public static <Item extends IItem> List<Item> getAllItems(FastAdapter<Item> fastAdapter) {
        int size = fastAdapter.getItemCount();
        List<Item> items = new ArrayList<>(size);
        for (int i = 0; i < size; i++) {
            Item item = fastAdapter.getItem(i);
            items.add(item);
            addAllSubItems(item, items);
        }
        return items;
    }

    /**
     * Gets all subItems from a given parent item
     *
     * @param item  the parent from which we add all items
     * @param items the list in which we add the subItems
     */
    public static <Item extends IItem> void addAllSubItems(Item item, List<Item> items) {
        if (item instanceof IExpandable && !((IExpandable) item).isExpanded() && ((IExpandable) item).getSubItems() != null) {
            List<Item> subItems = (List<Item>) ((IExpandable<Item, ?>) item).getSubItems();
            Item subItem;
            for (int i = 0, size = subItems.size(); i < size; i++) {
                subItem = subItems.get(i);
                items.add(subItem);
                addAllSubItems(subItem, items);
            }
        }
    }
}

File: library-extensions-expandable/src/main/java/com/mikepenz/fastadapter/utils/AdapterUtil.kt
package com.mikepenz.fastadapter.utils

import com.mikepenz.fastadapter.FastAdapter
import com.mikepenz.fastadapter.GenericItem
import com.mikepenz.fastadapter.IExpandable
import java.util.ArrayList

/**
 * Created by mikepenz on 31.12.15.
 */
object AdapterUtil {
    /**
     * Internal method to restore the selection state of subItems
     *
     * @param item          the parent item
     * @param selectedItems the list of selectedItems from the savedInstanceState
     */
    @JvmStatic fun <Item> restoreSubItemSelectionStatesForAlternativeStateManagement(item: Item, selectedItems: List<String>?) where Item : GenericItem, Item : IExpandable<*> {
        if (!item.isExpanded) {
            val subItems = (item as IExpandable<*>).subItems
            for (i in subItems.indices) {
                val subItem = subItems[i] as Item
                val id = subItem.identifier.toString()
                if (selectedItems != null && selectedItems.contains(id)) {
                    subItem.isSelected = true
                }
                restoreSubItemSelectionStatesForAlternativeStateManagement(subItem, selectedItems)
            }
        }
    }

    /**
     * Internal method to find all selections from subItems and sub sub items so we can save those inside our savedInstanceState
     *
     * @param item       the parent item
     * @param selections the ArrayList which will be stored in the savedInstanceState
     */
    @JvmStatic fun <Item> findSubItemSelections(item: Item, selections: MutableList<String>) where Item : GenericItem, Item : IExpandable<*> {
        if (!item.isExpanded) {
            val subItems = (item as IExpandable<*>).subItems
            for (i in subItems.indices) {
                val subItem = subItems[i] as Item
                val id = subItem.identifier.toString()
                if (subItem.isSelected) {
                    selections.add(id)
                }
                findSubItemSelections(subItem, selections)
            }
        }
    }

    /**
     * Gets all items (including sub items) from the FastAdapter
     *
     * @param fastAdapter the FastAdapter
     * @return a list of all items including the whole subItem hirachy
     */
    @JvmStatic fun <Item> getAllItems(fastAdapter: FastAdapter<Item>): List<Item> where Item : GenericItem, Item : IExpandable<*> {
        val size = fastAdapter.itemCount
        val items = ArrayList<Item>(size)
        for (i in 0 until size) {
            fastAdapter.getItem(i)?.let {
                items.add(it)
                addAllSubItems(it, items)
            }
        }
        return items
    }

    /**
     * Gets all subItems from a given parent item
     *
     * @param item  the parent from which we add all items
     * @param items the list in which we add the subItems
     */
    @JvmStatic fun <Item> addAllSubItems(item: Item?, items: MutableList<Item>) where Item : GenericItem, Item : IExpandable<*> {
        if (item is IExpandable<*> && !item.isExpanded) {
            val subItems = (item as IExpandable<*>).subItems
            var subItem: Item
            for (i in subItems.indices) {
                subItem = subItems[i] as Item
                items.add(subItem)
                addAllSubItems(subItem, items)
            }
        }
    }
}
